### PR TITLE
Prettier and eslint improvements (no dependency upgrades yet)

### PR DESCRIPTION
**Pull request by mattpocock**
- Make prettier run on the entire repo, not just the source files
- Remove prettier from eslint configs
- Made the CI check that code is formatted correctly

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -41,8 +42,6 @@
     "@typescript-eslint/parser": "4.28.1",
     "babel-jest": "27.5.1",
     "eslint": "7.30.0",
-    "eslint-config-prettier": "7.1.0",
```
[the remark this answers is not in the document]
Comment (mattpocock) [reply package.json:44]: I agree with this:

https://www.joshuakgoldberg.com/blog/you-probably-dont-need-eslint-config-prettier-or-eslint-plugin-prettier/

[the remark this answers is not in the document]
Comment (none23) [reply package.json:44]: Thanks for the link

**Review by none23 — APPROVED** [review]

File: package.json [hunk 2]
```diff
@@ -12,7 +12,8 @@
     "test": "jest && npm run test-types",
     "test-types": "tsc --noEmit -p ./tests/tsconfig.tests.json",
     "lint": "eslint ./src --ext .ts",
-    "format": "prettier --write 'src/**/*.ts?(x)' && npm run lint -- --fix",
+    "format": "prettier --write . && npm run lint -- --fix",
+    "format-check": "prettier --check .",
```
Comment (dmmulroy) [inline package.json:16]: Lets utilize prettier and eslint's cacheing options 

- https://prettier.io/docs/en/cli#--cache
- https://eslint.org/docs/latest/use/command-line-interface#caching

```suggestion
    "lint": "eslint ./src --ext .ts --cache --cache-location 'node_modules/.cache/eslintcache",
    "format": "prettier --write --cache --cache-location='node_modules/.cache/prettiercache' . && npm run lint -- --fix",
    "format-check": "prettier --check --cache --cache-location='node_modules/.cache/prettiercache' .",
```

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmmulroy) [reply package.json:44]: I also agree with removing `eslint-config-prettier` and the prettier docs advise to essentially do the same (i.e. let linters lint, and formatters format)

see: https://prettier.io/docs/en/comparison

**Review by dmmulroy — CHANGES_REQUESTED** [review]

**Review by supermacro — APPROVED** [review]
lgtm 👍 last things:

- let's just make sure we make the one change as per @dmmulroy's comment
  - https://github.com/supermacro/neverthrow/pull/574#discussion_r1747148379
- Resolve merge conflicts

File: package.json (hunk 2 above)
Comment (mattpocock) [reply package.json:16]: Getting this warning from Prettier:

```
[warn] Ignored unknown option --cache.
[warn] Ignored unknown option --cache-location=node_modules/.cache/prettiercache.
```

**Comment by mattpocock** [comment]
@supermacro The CI appears to be failing because a rollup dependency isn't being added. Have you seen this before? Unsure how to fix. The next solution I'd try would be to delete `package-lock.json` and remake it, but don't want to bloat the PR.

**Comment by mattpocock** [comment]
Might be something to do with the GH action cache?

File: tests/index.test.ts [hunk 3]
```diff
@@ -703,15 +699,15 @@ describe('Utils', () => {
...
+        type HeterogenousList = [
+          ResultAsync<string, string>,
+          ResultAsync<number, number>,
+          ResultAsync<boolean, boolean>,
         ]
 
-        type ExpecteResult = Result<[ string, number, boolean ], (string | number | boolean)[]>
+        const heterogenousList: HeterogenousList = [okAsync('Yooooo'), okAsync(123), okAsync(true)]
+
+        type ExpecteResult = Result<[string, number, boolean], [string, number, boolean]>
```
Comment (none23) [inline tests/index.test.ts:710]: Is this change intentional?

Comment (none23) [reply tests/index.test.ts:710]: @mattpocock Isn't that the reason tests are failing?

Comment (mattpocock) [reply tests/index.test.ts:710]: Apparently not intentional, let me check

Comment (mattpocock) [reply tests/index.test.ts]: Looks like I was pulling from origin/master instead of upstream/master, my bad. Fixed.

**Comment by mattpocock** [comment]
This PR is getting a bit out of hand. I am having real trouble rebasing every time a new change is merged. I'll create a new PR which adds the ability to run prettier, but doesn't format the whole repo. Once that's merged we can format the repo properly.
